File: src/main/java/vokorpgback/feature/character/domain/model/Abilities.java
package vokorpgback.feature.character.domain.model;

import vokorpgback.commons.Validation;
import vokorpgback.feature.commons.domain.model.GameMode;
import vokorpgback.feature.commons.domain.port.Dice;

public record Abilities(
                Ability strength,
                Ability agility,
                Ability perception) {

        public Abilities {
                Validation.require(strength.isPositive(), "Strength must be positive.");
                Validation.require(agility.isPositive(), "Agility must be positive.");
                Validation.require(perception.isPositive(), "Perception must be positive.");
        }

        public static Abilities generateAbilities(GameMode gameMode, Dice dice) {
                return new Abilities(
                                Ability.generateAbility(gameMode, dice),
                                Ability.generateAbility(gameMode, dice),
                                Ability.generateAbility(gameMode, dice));
        }
}

File: src/main/kotlin/core/domain/hero/Abilities.kt
package core.domain.hero

import core.domain.sharedkernel.Dice
import core.domain.sharedkernel.GameMode.EASY
import core.domain.sharedkernel.GameMode.NORMAL

data class Abilities(
    val strength: Ability,
    val agility: Ability,
    val perception: Ability
) {
    companion object {
        // TODO: naming
        fun randomInNormalMode() = Abilities(
            strength = Ability.randomInNormalMode(),
            agility = Ability.randomInNormalMode(),
            perception = Ability.randomInNormalMode()
        )

        // TODO: naming
        fun randomInEasyMode() = Abilities(
            strength = Ability.randomInEasyMode(),
            agility = Ability.randomInEasyMode(),
            perception = Ability.randomInEasyMode()
        )
    }

    // TODO: override operator plus ?
    fun sum() = strength.value + agility.value + perception.value
}

@JvmInline
value class Ability(val value: Int) {
    companion object {
        private val dice = Dice

        // TODO: naming
        fun randomInNormalMode() = Ability(dice.sixSidedRoll() + NORMAL.abilityModifier)
        fun randomInEasyMode() = Ability(dice.sixSidedRoll() + EASY.abilityModifier)
    }

    init {
        require(value > 0) { "An ability should be positive. Value = $value." }
    }

    operator fun plus(other: Int) = value + other
}